### PR TITLE
backupccl: remap all restored tables

### PR DESCRIPTION
**Pull request by dt**
This PR has a few changes, broken down into separate commits:
a) stop restoring tmp tables and remove the special-case code to synthesize their special schemas; These were previously restored only to be dropped so that restored jobs that referenced them would not be broken, but we stopped restoring jobs.
b) synthesize type-change jobs during cluster restore; this goes with not restoring jobs.
c) fix some assumptions in tests/other code about what IDs restored tables have.
d) finally, always assign new IDs to all restored objects, even during cluster restore, removing the need to carefully move conflicting tables or other things around.

Commit-by-commit review recommended.

### CODE REVIEW

**Review by benbardin — APPROVED** [review]
LGTM, but consider also waiting for another review or two from the others :)

**Review by ajwerner — APPROVED** [review]
:lgtm: though the test failure seems plausibly legit. This was satisfying to read. 

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 1 of 3 files at r5, 4 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85492#-:-N8ZY7juBo3JHdmO-AQR:b-dngmjt)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru and @dt)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85492#-N8ZR2YZAywAtTmDwUZu):*
I'm so curious
___
*[`pkg/ccl/backupccl/restore_job.go` line 1667 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85492#-N8ZX2HNA3B18t2hRfKj:-N8ZX2HNA3B18t2hRfKk:bggeg8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f71d99b233bc547a40e91f1791609a247439d8f/pkg/ccl/backupccl/restore_job.go#L1667)):*
> ```Go
> 
> // tempSystemDatabaseID returns the ID of the descriptor for the temporary
> // system database used in full cluster restores.
> ```

Can you note that this returns `descpb.InvalidID` in the case that there are no system tables in the `tables` slice and that it relies on the `DescriptorRewrites` map having already been populated?
___
*[`pkg/ccl/backupccl/restore_old_versions_test.go` line 543 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85492#-N8ZXVjI1-oNp2IoDp09:-N8ZXVjI1-oNp2IoDp0A:bk00v3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/72c42549c47c01cab84a3991d4b3450bb82de5dc/pkg/ccl/backupccl/restore_old_versions_test.go#L543)):*
> ```Go
> 			{"root", "", "{admin}"},
> 		})
> 		sqlDB.CheckQueryResults(t, "SELECT comment FROM system.comments", [][]string{
> ```

technically, should this have an order by?

File: pkg/ccl/backupccl/restore_job.go
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_job.go:1666]: nit: can you add some more words tying this to the logic in planning. It will help future readers to understand why its okay to just use the first system tables parentID rewrite.

File: pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts
Comment (adityamaru) [inline pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts:84]: any reason to keep this statement around? Seems redundant with the previous one.

File: pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts
Comment (adityamaru) [inline pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts:156]: hmm why don't we restore this anymore?

Comment (stevendanna) [inline pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts:156]: Is this expected?

**Review by adityamaru — CHANGES_REQUESTED** [review]
Can I bug you to flesh out the commit message in commit number 5. For anyone who comes along later it would be helpful to read why we made this switch etc. Feel free to ignore.

**Review by stevendanna — COMMENTED** [review]
Overall, it is nice to see this code get deleted.

One question I have is why we don't need the custom restore funcs for comments, session vars, and zone configuration all of which reference descriptor IDs in their table.  If we _do_ need them, then perhaps the descriptor-conflicts tests isn't testing exactly what it used to be.

File: pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts
Comment (dt) [reply pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts:84]: I figured we want to test both directions on the namespace table since the conflicts bug created duplicate entries and all

**Comment by dt** [comment]
@stevendanna 
> One question I have is why we don't need the custom restore funcs for comments, session vars, and zone configuration all of which reference descriptor IDs in their table.
This is why https://github.com/cockroachdb/cockroach/pull/85098/files landed first

**Comment by stevendanna** [comment]
@dt Ah, thanks. missed that one.

File: pkg/ccl/backupccl/testdata/backup-restore/temp-tables
Comment (dt) [reply pkg/ccl/backupccl/testdata/backup-restore/temp-tables:155]: It seemed like it was showing that we didn't have a tmp table post-restore, so that was worth keeping, but I could be convinced it isn't. It can also go away later?

File: pkg/ccl/backupccl/testdata/backup-restore/temp-tables
Comment (dt) [reply pkg/ccl/backupccl/testdata/backup-restore/temp-tables:135]: Done.

File: pkg/ccl/backupccl/testdata/backup-restore/temp-tables
Comment (dt) [reply pkg/ccl/backupccl/testdata/backup-restore/temp-tables]: Done.

File: pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts
Comment (dt) [reply pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts]: Ha, nice catch -- there was a bug in the prior PR that this was pointing to that was previously hidden by the fact we remapped 1:1. Fixed with new first commit here.

Comment (dt) [reply pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts]: No, see reply to Aditya above.

Comment (dt) [reply pkg/ccl/backupccl/testdata/backup-restore/descriptor-conflicts]: er, by first I meant 5th: [backupccl: remap all IDs in one pass](https://github.com/cockroachdb/cockroach/pull/85492/commits/6c54e9a8e6430c0f3b496d0909ae86f10920d50d)

**Comment by dt** [comment]
> Can I bug you to flesh out the commit message in commit number 5. For anyone who comes along later it would be helpful to read why we made this switch etc. Feel free to ignore.

Added a lengthy commit message.

File: pkg/ccl/backupccl/system_schema.go
Comment (adityamaru) [inline pkg/ccl/backupccl/system_schema.go:489]: detritus?

**Review by adityamaru — APPROVED** [review]

**Comment by dt** [comment]
TFTRs!

bors r+

**Pull request merged**
